basic support for --keep-git-dir, --unpack, --parents

aokellermann opened the pull request:
`ADD`:
- [--keep-git-dir](https://docs.docker.com/reference/dockerfile/#add---keep-git-dir)
- [--unpack](https://docs.docker.com/reference/dockerfile/#add---unpack)

`COPY`:
- [--parents](https://docs.docker.com/reference/dockerfile/#copy---parents)

aokellermann committed:
add keep-git-dir, unpack, parents

rene commented:
Dear maintainers, this issue has been impacting [EVE-OS](https://github.com/lf-edge/eve) project for a while, any chances to get this PR merged?

https://github.com/lf-edge/eve/pull/5839/changes#diff-388d69413dd217e9bd38afa91aa10c32a35fa8f76f91b8b71e85eb8d7116b0c0R18

aokellermann commented:
just rebased with Claude. 

pinging @m-ildefons

m-ildefons commented on src/Language/Docker/Parser/Copy.hs:
```diff
@@ -159,6 +179,16 @@ link = do
+parents :: Parser Parents
+parents = do
+  void $ string "--parents"
+  return Parents
```
This flag optionally takes an explicit boolean value, i.e. the following three variations are all valid syntax:
```
COPY --parents foo bar
```
```
COPY --parents=true foo bar
```
```
COPY --parents=false foo bar
```
This parser only supports the first case, but would fail to correctly work for the other two cases.

https://docs.docker.com/reference/dockerfile/#copy---parents

m-ildefons commented on src/Language/Docker/Parser/Copy.hs:
```diff
@@ -159,6 +179,16 @@ link = do
+keepGitDir :: Parser KeepGitDir
+keepGitDir = do
+  void $ string "--keep-git-dir"
+  return KeepGitDir
```
This flag takes an obligatory boolean value, much like the `--unpack` flag.
This parser doesn't accept that.

https://docs.docker.com/reference/dockerfile/#add---unpack

m-ildefons commented on src/Language/Docker/PrettyPrint.hs:
```diff
@@ -160,6 +160,18 @@ prettyPrintLink link =
     Link -> "--link"
     NoLink -> mempty
 
+prettyPrintKeepGitDir :: KeepGitDir -> Doc ann
+prettyPrintKeepGitDir keepGitDir =
+  case keepGitDir of
+    KeepGitDir -> "--keep-git-dir"
+    NoKeepGitDir -> mempty
```
This should either produce `--keep-git-dir=true`, `--keep-git-dir=false` or nothing, since the boolean value for the flag is not optional.

m-ildefons commented on src/Language/Docker/Syntax.hs:
```diff
@@ -162,6 +162,16 @@ data Link
+  | NoKeepGitDir
+  deriving (Show, Eq, Ord)
+
+data Parents
+  = Parents
+  | NoParents
+  deriving (Show, Eq, Ord)
```
This datastructure doesn't accurately model the `--parents` flag. It should be more like the data structure `Unpack` below, because that allows a distinction between the absence of the flag and an explicit `--parents=false`. The distinction is important to be able to make the pretty printer either omit the flag, or produce an explicit `--parents=false`.

m-ildefons commented on test/Language/Docker/ParseAddSpec.hs:
```diff
@@ -47,47 +47,55 @@ spec = do
+            file
+            [ Add
+                ( AddArgs (fmap SourcePath ["foo"]) (TargetPath "bar") )
+                ( AddFlags NoChecksum NoChown NoChmod NoLink KeepGitDir NoUnpack [] )
```
Missing test cases for `--keep-git-dir=true` and `--keep-git-dir=false`

m-ildefons commented on test/Language/Docker/ParseCopySpec.hs:
```diff
@@ -68,36 +68,44 @@ spec = do
+            file
+            [ Copy
+                ( CopyArgs [ SourcePath "source" ] ( TargetPath "/target" ) )
+                ( CopyFlags NoChown NoChmod NoLink Parents NoSource [])
```
Missing test cases for `--parents=true` and `--parents=false`

m-ildefons commented on test/Language/Docker/PrettyPrintSpec.hs:
```diff
@@ -24,57 +24,63 @@ spec = do
+      let add = Add
+                  ( AddArgs [SourcePath "foo"] (TargetPath "bar") )
+                  ( AddFlags NoChecksum NoChown NoChmod NoLink KeepGitDir NoUnpack [] )
+       in assertPretty "ADD --keep-git-dir foo bar" add
```
Missing test cases for --keep-git-dir=true and --keep-git-dir=false

m-ildefons commented on test/Language/Docker/PrettyPrintSpec.hs:
```diff
@@ -86,30 +92,35 @@ spec = do
+      let copy = Copy
+                  ( CopyArgs [SourcePath "foo"] (TargetPath "bar") )
+                  ( CopyFlags NoChown NoChmod NoLink Parents NoSource [] )
+       in assertPretty "COPY --parents foo bar" copy
```
Missing test cases for --parents=true and --parents=false

m-ildefons reviewed (commented):
Thanks for the PR.
It's just not quite there yet: there are missing test cases and the parser doesn't really model what the syntax reference specifies for the two flags.

m-ildefons committed:
Add explicit/implicit versions for flags

Add explicit and implicit versions for --unpack, --parents and
--keep-git-dir flags.
This means that

COPY --unpack foo bar

and

COPY --unpack=true foo bar

are now equivalent and both valid.
Similarly for --parents and --keep-git-dir flags as well.

Signed-off-by: Moritz Röhrich <moritz@ildefons.de>

m-ildefons commented:
Again, thanks for the PR. I've gone ahead and fixed it up a bit, so it supports both the explicit and the implicit versions of these flags, i.e. `--parents` and `--parents=true`.
